### ISSUE TITLE
iTunes library metadata and smart playlists

### ISSUE THREAD

**Issue by jptrsn** [ejurgensen/forked-daapd#549]
I'm in the process of porting over my smart playlist logic into the correct format for forked-daapd. I'm using the ratings as well as last played dates to calculate what songs should be in my list, but I'm using the iTunes library XML metadata, as I don't want to go through the process of rating 18 Gb of music again.

It looks like the logic rules are not using the metadata pulled from the iTunes XML, although I do have that option set in the config file.

I've got a simple list that I believe confirms my suspicion
```
"Five Stars" {
  rating = 5
}
```
Is there any way to clone this metadata into the local database, or to force the smart playlists to use the imported metadata? I'd like to keep all metadata in one place so I don't need to maintain two separate databases.

**Comment by ejurgensen** [comment]
You did the right thing and it should work :-)

Since it didn't, either the XML import of the rating isn't working, or the smart playlist isn't making the correct query. I will see if I can reproduce. If you are able to do a bit of SQL, you can help yourself. Run sqlite3 with songs3.db, and then select rating and path from the files table. If it is incorrect, then the issue is with the XML import.

**Comment by jptrsn** [comment]
That was the clue I needed! Looking at the records, I realized that in the database, records are a multiple of 20, up to 100, so my query was working as expected. However, it appears that `time_played` and `play_count` aren't coming in as it shows in iTunes.

I'm a bit of a metadata hound - I use it to compile some pretty cool queries for my shuffled playlist. Is there any chance of adding `skip_count` and `time_skipped`, and potentially the ability to include or exclude based on presence in another playlist? I use that to sort more deeply, and manage more subjective variables in my overall system.

Here's the basic implementation, but without the logic to account for skips, date added, and the union and exclusion of a couple of manual playlists. Notice that I exclude all 1 star songs, which is what I use to tag songs I don't want to hear in a mix, but want to keep as part of an album

```
"Smart Shuffle" {
media_kind is music
  and not (genre is "Folk" or genre is "World" or genre is "Holiday" or genre is "Children" or genre is "Speech" or genre is "Humor" or genre is "Jazz")
  and not (rating = 20)
  and (
      (rating = 100 and
        (
          (play_count = 0)
          or (play_count < 10 and time_played before 8 days ago)
          or (play_count > 9 and play_count < 20 and time_played before 13 days ago)
          or (play_count > 19 and play_count < 30 and time_played before 18 days ago)
          or (play_count > 29 and time_played before 30 days ago)
        )
      )
    or
      (rating = 80 and
        (
          (play_count = 0)
          or (play_count < 10 and time_played before 13 days ago)
          or (play_count > 9 and play_count < 20 and time_played before 18 days ago)
          or (play_count > 19 and play_count < 30 and time_played before 23 days ago)
          or (play_count > 29 and time_played before 60 days ago)
        )
      )
    or
      (rating = 60 and
        (
          (play_count = 0)
          or (play_count < 10 and time_played before 18 days ago)
          or (play_count > 9 and play_count < 20 and time_played before 23 days ago)
          or (play_count > 19 and play_count < 30 and time_played before 28 days ago)
          or (play_count > 29 and time_played before 90 days ago)
        )
      )
    or
      (rating = 60 and
        (
          (play_count = 0)
          or (play_count < 10 and time_played before 18 days ago)
          or (play_count > 9 and play_count < 20 and time_played before 23 days ago)
          or (play_count > 19 and play_count < 30 and time_played before 28 days ago)
          or (play_count > 29 and time_played before 90 days ago)
      )
    )
  )
}

```

**Comment by ejurgensen** [comment]
I don't think forked-daapd multiplies the rating, so it must come multiplied from iTunes. Is time_played and play_count exported by iTunes in the XML? I don't see it in my XML's.

Adding skip_count and time_skipped as file metadata is a bit of work, and not on the top of my list.

I see the idea behind the smart playlist, but I would think that it transforms into a monster of a SQL query that might take a while on your 18GB of music? If that is the case you could consider a nightly cron job which updates/recreates a m3u playlist.

**Comment by jptrsn** [comment]
I see the idea behind updating the playlist. Both play date and count is in the xml, as well as skip count and timestamp values. Would the code handle extra columns in the files table, if I were to inject a few to include this data? I'd then be able to manage the playlist with a scheduled task, as you suggested.

Here's a snip from my xml iTunes export.

```
<key>615</key>
		<dict>
			<key>Track ID</key><integer>615</integer>
			<key>Size</key><integer>7602660</integer>
			<key>Total Time</key><integer>317178</integer>
			<key>Track Number</key><integer>4</integer>
			<key>Year</key><integer>1975</integer>
			<key>BPM</key><integer>122</integer>
			<key>Date Modified</key><date>2016-08-09T15:07:35Z</date>
			<key>Date Added</key><date>2009-03-27T15:37:39Z</date>
			<key>Bit Rate</key><integer>192</integer>
			<key>Sample Rate</key><integer>44100</integer>
			<key>Volume Adjustment</key><integer>255</integer>
			<key>Play Count</key><integer>94</integer>
			<key>Play Date</key><integer>3598042339</integer>
			<key>Play Date UTC</key><date>2018-01-06T05:12:19Z</date>
			<key>Skip Count</key><integer>17</integer>
			<key>Skip Date</key><date>2018-02-28T22:17:40Z</date>
			<key>Rating</key><integer>80</integer>
			<key>Album Rating</key><integer>60</integer>
			<key>Persistent ID</key><string>735C22B6342B6B74</string>
			<key>Track Type</key><string>File</string>
			<key>File Folder Count</key><integer>-1</integer>
			<key>Library Folder Count</key><integer>-1</integer>
			<key>Name</key><string>Wish You Were Here</string>
			<key>Artist</key><string>Pink Floyd</string>
			<key>Album Artist</key><string>Pink Floyd</string>
			<key>Composer</key><string>David Gilmour/Roger Waters</string>
			<key>Album</key><string>Wish You Were Here</string>
			<key>Genre</key><string>Classic Rock</string>
			<key>Kind</key><string>MPEG audio file</string>
			<key>Equalizer</key><string>#!#116#!#</string>
			<key>Equalizer</key><string>Rock</string>
			<key>Sort Album</key><string>Wish You Were Here</string>
			<key>Location</key><string>file://localhost/E:/Music/Music/Pink%20Floyd/Wish%20You%20Were%20Here/04%20Wish%20You%20Were%20Here.mp3</string>
		</dict>
```

**Comment by ejurgensen** [comment]
I've added mapping of Play Date and Play Count.

I suppose you could add columns for skip, it wouldn't break anything ... right now. Later it might be a problem if forked-daapd gets more columns. You would also have a challenge on how to map the iTunes id with forked-daapd's id. There is a bunch of logic to do that in forked-daapd that you would need to copy.

If you are up to it, and know C, you could try making the change and then submit it as a PR? But be prepared that I may ask you to make small or large adjustments before including it.

**Comment by jptrsn** [comment]
I'm a pretty horrible coder in C, but I'll take a crack at it. Essentially, I'll just clone what you've done for play_count and last_played, replicated for skips. Think that will work?

**Comment by ejurgensen** [comment]
If it was just that I would have done it :-)

If you are not a C coder maybe better if I take a look at the task beforehand.

**Comment by ejurgensen** [comment]
I took a look at it, and maybe I overstated how much was required, in any case I ended up making the changes I think are required. Right now they are in this branch:

https://github.com/ejurgensen/forked-daapd/tree/skip_count

**Comment by jptrsn** [comment]
Amazing! I guess I need to follow the build instructions to pull that branch and build for my pi zero w?

**Comment by ejurgensen** [comment]
Yes

**Comment by ejurgensen** [comment]
This skip_count branch is now merged into the master branch

**Comment by jptrsn** [comment]
Things are working a treat, thanks for including those fields!

A quick follow-up question: is there a way to import iTunes metadata only once? I'd like to migrate my library away from iTunes and use forked-daapd to track and manage metadata like ratings and play counts.

**Comment by ejurgensen** [comment]
>is there a way to import iTunes metadata only once?

Can you elaborate? I don't quite understand the question... you decide yourself how often you want to import from iTunes.

**Comment by jptrsn** [comment]
I guess I do not understand the functionality then. How would I get all my iTunes metadata into the songs database, but then be able to modify it from forked-daapd only? If I import from iTunes, then modify the conf file to disable iTunes metadata override, will that leave the original metadata in the database?

**Comment by ejurgensen** [comment]
For regular files forked-daapd can import metadata from two sources, either the file metadata tags, or iTunes xml (well some metadata can also be set via clients, e.g. ratings and play counts, but that is probably not relevant for your question).

The setting in the conf file tells forked-daapd if metadata from the iTunes xml should have priority over the file id3 tags.

If you afterwards delete the xml or disable the setting, the metadata will still be there. However, if a file later gets modified then forked-daapd will notice it and use the latest file metadata.

**Comment by ejurgensen** [comment]
The content of a smart playlists is determined when you make the request, so there is no recalculation or caching. I suggest you open a new issue about last_played/last_skipped where you give an example of what you observe, and what you expect. Also try setting the log level to debug, so you can include the query that is made.

**Comment by chme** [comment]
As far as i can remember, smart playlist may get cached for daap/dacp clients (depending on how long it takes to process the daap/dacp request to load the smart playlist). This means daap/dacp clients will only get the refreshed smart playlist after updating the library (updating play/skip count does not trigger rebuilding the daap/dacp cache).

**Comment by ejurgensen** [comment]
Yes seems I was wrong, I thought smart playlists were excluded from caching, but it doesn’t look like it. Perhaps they should be, not sure about that.

**Comment by jptrsn** [comment]
Turns out it was Chrome browser caching the playlist. When I force-refresh, it updates the playlist properly and I can see the query in the logs. I couldn't seem to find a setting to force the browser to refresh the page contents every time, but a small change to the page source should force it to update. If I understood python and knew how to build and test, I'd submit a pull request.

**Comment by chme** [comment]
The player web interface adds a cache control header allowing browsers to cache the json replies. This mimics the caching for daapd clients.

The relevant lines:
https://github.com/ejurgensen/forked-daapd/blob/4cf71bafcc8311896f62c8ba1eb8255a2f73d671/src/httpd_jsonapi.c#L2391-L2393

@ejurgensen it is possible to disable caching for the playlist json api endpoint. But this would result in longer loading times for unchanged playlists. Not sure what the best way to deal with this is ...  Do you have a preference?

**Comment by ejurgensen** [comment]
Yes, the old problem of cache invalidation. I'm not thrilled about just disabling, as you say it will slow things down (same goes for the daap cache). Can we think of a better solution? Maybe mark specific smart playlists as none-cachable, i.e. if they use playcount/skipcount?

**Comment by chme** [comment]
Instead of using the DB_ADMIN_UPDATE timestamp, we could store a new timestamp in the playlist table. This timestamp can be used for the "not modified since" header and would be updated during inc_playcount/skipcount for all playlist that contain playcount/skipcount/time_played/time_skipped. 
(I thought about using the db_timestamp, but updating it outside of the filescan might lead to problems if inotify is not used)
